File: server/paper/src/main/java/su/plo/lib/paper/entity/PaperServerEntity.java
package su.plo.lib.paper.entity;

import lombok.RequiredArgsConstructor;
import org.bukkit.Location;
import org.bukkit.entity.LivingEntity;
import org.bukkit.util.Vector;
import org.jetbrains.annotations.NotNull;
import su.plo.lib.api.entity.MinecraftPlayer;
import su.plo.lib.api.server.MinecraftServerLib;
import su.plo.lib.api.server.entity.MinecraftServerEntity;
import su.plo.lib.api.server.world.MinecraftServerWorld;
import su.plo.lib.api.server.world.ServerPos3d;
import su.plo.voice.proto.data.pos.Pos3d;

import java.util.UUID;

@RequiredArgsConstructor
public class PaperServerEntity<E extends LivingEntity> implements MinecraftServerEntity {

    protected final MinecraftServerLib minecraftServer;
    protected final E instance;

    private final Pos3d position = new Pos3d();
    private final Pos3d lookAngle = new Pos3d();

    private Location location;

    @Override
    public int getId() {
        return instance.getEntityId();
    }

    @Override
    public @NotNull UUID getUUID() {
        return instance.getUniqueId();
    }

    @Override
    public @NotNull Pos3d getPosition() {
        return getPosition(position);
    }

    @Override
    public @NotNull Pos3d getPosition(@NotNull Pos3d position) {
        fetchLocation();

        position.setX(location.getX());
        position.setY(location.getY());
        position.setZ(location.getZ());

        return position;
    }

    @Override
    public @NotNull Pos3d getLookAngle() {
        return getLookAngle(lookAngle);
    }

    @Override
    public @NotNull Pos3d getLookAngle(@NotNull Pos3d lookAngle) {
        Vector vector = instance.getLocation().getDirection();
        lookAngle.setX(vector.getX());
        lookAngle.setY(vector.getY());
        lookAngle.setZ(vector.getZ());

        return lookAngle;
    }

    @Override
    public double getEyeHeight() {
        return instance.getEyeHeight();
    }

    @Override
    public float getHitBoxWidth() {
        return (float) instance.getBoundingBox().getWidthX();
    }

    @Override
    public float getHitBoxHeight() {
        return (float) instance.getBoundingBox().getHeight();
    }

    @Override
    public boolean isInvisibleTo(@NotNull MinecraftPlayer player) {
        throw new IllegalStateException("Not implemented"); // todo: implement
    }

    @Override
    public <T> T getInstance() {
        return (T) instance;
    }

    @Override
    public @NotNull ServerPos3d getServerPosition() {
        fetchLocation();

        return new ServerPos3d(
                minecraftServer.getWorld(instance.getWorld()),
                location.getX(),
                location.getY(),
                location.getZ(),
                location.getYaw(),
                location.getPitch()
        );
    }

    @Override
    public @NotNull ServerPos3d getServerPosition(@NotNull ServerPos3d position) {
        fetchLocation();

        position.setWorld(minecraftServer.getWorld(instance.getWorld()));

        position.setX(location.getX());
        position.setY(location.getY());
        position.setZ(location.getZ());

        position.setYaw(location.getYaw());
        position.setPitch(location.getPitch());

        return position;
    }

    @Override
    public @NotNull MinecraftServerWorld getWorld() {
        return minecraftServer.getWorld(instance.getWorld());
    }

    private void fetchLocation() {
        if (location == null) {
            this.location = instance.getLocation();
        } else {
            instance.getLocation(location);
        }
    }
}

File: server/paper/src/main/kotlin/su/plo/lib/paper/entity/PaperServerEntity.kt
package su.plo.lib.paper.entity

import lombok.RequiredArgsConstructor
import org.bukkit.Location
import org.bukkit.entity.LivingEntity
import org.bukkit.entity.Player
import su.plo.lib.api.entity.MinecraftPlayerEntity
import su.plo.lib.api.server.MinecraftServerLib
import su.plo.lib.api.server.entity.MinecraftServerEntity
import su.plo.lib.api.server.world.ServerPos3d
import su.plo.voice.proto.data.pos.Pos3d

@RequiredArgsConstructor
open class PaperServerEntity<E : LivingEntity>(
    protected val minecraftServer: MinecraftServerLib,
    protected val instance: E
) : MinecraftServerEntity {

    private val position = Pos3d()
    private val lookAngle = Pos3d()

    private var location: Location? = null

    override fun getId() = instance.entityId

    override fun getUUID() = instance.uniqueId

    override fun getPosition() = getPosition(position)

    override fun getPosition(position: Pos3d): Pos3d {
        val location = fetchLocation()

        position.x = location.x
        position.y = location.y
        position.z = location.z

        return position
    }

    override fun getLookAngle() = getLookAngle(lookAngle)

    override fun getLookAngle(lookAngle: Pos3d): Pos3d {
        val vector = instance.location.direction

        lookAngle.x = vector.x
        lookAngle.y = vector.y
        lookAngle.z = vector.z

        return lookAngle
    }

    override fun getEyeHeight() = instance.eyeHeight

    override fun getHitBoxWidth() = instance.boundingBox.widthX.toFloat()

    override fun getHitBoxHeight() = instance.boundingBox.height.toFloat()

    override fun <T> getInstance() = instance as T

    override fun getServerPosition(): ServerPos3d {
        val location = fetchLocation()

        return ServerPos3d(
            minecraftServer.getWorld(instance.world),
            location.x,
            location.y,
            location.z,
            location.yaw,
            location.pitch
        )
    }

    override fun getServerPosition(position: ServerPos3d): ServerPos3d {
        val location = fetchLocation()

        position.world = minecraftServer.getWorld(instance.world)
        position.x = location.x
        position.y = location.y
        position.z = location.z
        position.yaw = location.yaw
        position.pitch = location.pitch

        return position
    }

    override fun getWorld() = minecraftServer.getWorld(instance.world)

    private fun fetchLocation(): Location {
        if (location == null) {
            location = instance.location
        } else {
            instance.getLocation(location)
        }

        return location!!
    }
}
